active/0021-transfer-files-over-mqtt.md

```diff
@@ -0,0 +1,205 @@
+
+* FTP and HTTP servers usually struggle to keep up with large number of simultaneous bandwidth-intensive connections
+* packet loss or reconnect forces clients to restart the transfer
+* devices which already talk MQTT need to integrate with one more SDK, address authenticaion and authorization, and potentially go through an additional round of security audit
+
+Known cases of device-to-cloud file transfer:
+
+* [CAN bus](https://en.wikipedia.org/wiki/CAN_bus) data
+* Image taken by industry camera for Quality Assurance
+* Large data file collected from forklift
+* Video and audio data from truck cars, and video data captured by inbound unloading cameras
+* Vehicle real-time logging, telemetry, messaging
+* Upload collected ML logs
+
+Known cases of cloud-to-device file transfer:
+
+* Upload AI/ML models
+* Firmware upgrades
+
+## Terminology
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [BCP 14](https://www.rfc-editor.org/bcp/bcp14) [[RFC2119](https://www.rfc-editor.org/rfc/rfc2119)] [[RFC8174](https://www.rfc-editor.org/rfc/rfc8174)] when, and only when, they appear in all capitals, as shown here.
+
+The following terms are used as described in [MQTT Version 5.0 Specification](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc39010030):
+* Application Message
+* Server
+* Client
+* Topic Name
+* Topic Filter
+* MQTT Control Packet
+
+*At least once*: a message can be delivered many times, but cannot be lost
+
+## Requirements
+
+* The protocol MUST use only PUBLISH type of MQTT Control Packet
+* The protocol MUST support transfer of file segments
+* Server MUST be able to verify integrity of each file segment
+* Client MAY know total file size when initiating the transfer
+* Client MAY abort file transfer
+* Server MAY ask the client to pause file transfer
+* Server MAY ask the client to abort file transfer
+* The protocol MUST NOT require changes in client code
+* The protocol MUST guarantee "At least once" delivery
+* Server MUST NOT support subscription on topics dedicated for file transfer
+
+## AWS IoT MQTT-based file delivery (reference design)
+
+As an example of existing implementation we can look at AWS IoT Core [which provides functionality](https://docs.aws.amazon.com/iot/latest/developerguide/mqtt-based-file-delivery.html) to [deliver files to IoT devices](https://docs.aws.amazon.com/iot/latest/developerguide/mqtt-based-file-delivery-in-devices.html):
+
+![](0021-assets/aws-mqtt-file-delivery.png)
+
+## Design
+
+### Overview
+
+* Files are split in segments of equal length with the exception of the last segment, it's length is > 0 and  <= segment length
+* Client generates UUID for each file being transferred and use it as file Id in Topic Name
```

Comment:
Not sure it's the right approach.
My thinking was as follows:
* we need some kind of unique file identifier anyway
* file path + file name could be fairly large, sending them as part of PUBLISH message with each segment could be greedy
* UUID can be stored in the session and broker can use it to get access to the rest of file metadata